base/reduce.jl

```diff
@@ -724,19 +724,29 @@ Returns the largest element in a collection.
 The value returned for empty `itr` can be specified by `init`. It must be
 a neutral element for `max` (i.e. which is less than or equal to any
 other element) as it is unspecified whether `init` is used
-for non-empty collections.
+for non-empty collections. If `init > maximum(itr)`, return `init`.
```

Comment:
I am okay with making the behavior specified here (c.f. https://github.com/JuliaLang/julia/issues/49042), but the docstring needs to be consistent and not both say it is unspecified behavior and to also specify the behavior here

Reply:
My understanding is that the current `maximum` (and `minimum` and maybe elsewhere?) documentation is merely repeating what is said about the `init` parameter of `mapreduce`. But this advice is only true for `mapreduce` in general, and not for the specific cases of `maximum` and `minimum`. Here we should be able to say `init` will indeed be treated as just another element in the input, and will be returned as the output in case of an empty list. It would just be nice to have 1. a confirmation that this is the desired behavior for `maximum`, `minimum` and `mapreduce` and 2. perhaps the ability to prove this is the case looking at the implementation, and what methods are called in the specific case of `maximum` and `minimum`. I'm unfortunately not too familiar with the implementation, I can't easily make sense of it myself, and I'm not sure where the implementation for these methods diverges compared to other reducing operations (if it diverges at all. _is_ it still unspecified in general for mapreduce?).

In other words, I suggest actually removing the text mentioning anything unspecified for these methods, and only mention the term will be output for an empty list, and will generally be treated as an extra item in the input.

Reply:
That behavior is already specified for mapfoldr, mapfoldl, and mapreduce(identity), so it seems reasonable to assume the same for mapreduce(max) and thus maximum as well. But that is the question intended to be answered by #49042 (it looks only like a doc change to me now).

Aside, to be pedantic about this question:

> reduce(min, [1,2,3],init=3) sort of follows the monoid laws as well... There's probably a special name for this group

I think this is exactly the same monoid law as the first example. In particular, the `init` is supposed to be ranging over the _domain_ of the inputs. So if the input was `UInt8` instead, then the init is `0xff` instead. But if the input function generating that array was `2pi*sin%Int`, then the init is arguably `6`, since the domain of that input function is `[-6, 6]`. Using `typemax` is just a rough approximation of the expected domain in any case.